README.md
# xapea00x

A Linux kernel driver for the Xaptum ENF Access card, a TPM 2.0-based
hardware card used to authenticate IoT devices and gateways to the
Xaptum Edge Network Fabric.

## Supported Product IDs

| Model # | Form Factor | USB Vendor ID | USB Product ID |
|---------|-------------|---------------|----------------|
| XAP-EA-001 | USB | 0x10C4 | 0x8BDE |
| XAP-EA-002 | USB | 0x2FE0 | 0x8BDE |
| XAP-EA-003 | Mini PCI-e | 0x2FE0 | 0x8BEE |
| XAP-EA-004 | MultiTech MTAC | 0x2FE0 | 0x8BFE |

## Runtime Dependencies

The necessary TPM 2.0 support was added to the 4.9 kernel, so this is
the earliest supported version.

The kernel must have been compiled with support for USB
(`USB_SUPPORT=y`), SPI (`SPI=y`), and TPM 2.0 (`TCG_TPM=y|m` and
`TCG_TIS_CORE=y|m`).

The `TCG_TIS_SPI` kernel module is also required.  Most Linux
distributions do not include it, so it must be built separately.
Further, kernels 4.9 through 4.11 contain a buggy version anyway.

This repo contains the source for a working version of `TCG_TIS_SPI`
to make out-of-tree building easy.

## Installation

### Debian (Stretch)

``` bash
# Install the Xaptum API repo GPG signing key.
apt-key adv --keyserver keyserver.ubuntu.com --recv-keys c615bfaa7fe1b4ca

# Add the repository to your APT sources.
echo "deb http://dl.bintray.com/xaptum/deb stretch main" > /etc/apt/sources.list.d/xaptum.list

# Install the library
sudo apt-get install xapea00x-dkms
```

### From Source

#### Build Dependencies

* `make`
* `gcc`
* Linux kernel headers for your distribution

#### xapea00x

```bash
make
sudo make modules_install
```

#### TCG_TIS_SPI (if needed)

If your system does not include a working `TCG_TIS_SPI` module, build
and install the copy in this repo:

```bash
cd tpm_tis_spi
make
sudo make modules_install
```

## License
Copyright (c) 2017-2020 Xaptum, Inc.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.


debian/changelog
xapea00x (0.1.3-1) bionic; urgency=medium

  * ubuntu: first version for bionic

 -- David R. Bild <david.bild@xaptum.com>  Fri, 21 May 2021 05:07:07 +0000

xapea00x (0.1.2-3) unstable; urgency=medium

  [ David R. Bild ]
  * Configure modules to use TPM as hardware rng

 -- Zane Beckwith <zane.beckwith@xaptum.com>  Wed, 11 Nov 2020 21:03:08 +0000

xapea00x (0.1.1-1) unstable; urgency=medium

  *  Fix USB timeout errors.

 -- David R. Bild <david.bild@xaptum.com>  Thu, 24 May 2018 14:11:28 +0000

xapea00x (0.1.0-1) unstable; urgency=medium

  * Initial release.

 -- David R. Bild <david.bild@xaptum.com>  Wed, 02 May 2018 16:42:23 -0500


xapea00x-spi.c
/* XAP-EA-00x driver for Linux
 *
 *  Copyright (c) 2017-2018 Xaptum, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#include "xapea00x.h"

/*******************************************************************************
 * SPI master functions
 */

/**
 * xapea00x_spi_setup - Setup the SPI channel for the TPM.
 *
 * Context: !in_interrupt()
 *
 * Return: If successful, 0. Otherwise a negative error number.
 */
static int xapea00x_spi_setup(struct spi_device *spi)
{
	struct xapea00x_device *dev;
	int retval;

	dev = spi_master_get_devdata(spi->master);

	mutex_lock(&dev->usb_mutex);
	if (!dev->interface) {
		retval = -ENODEV;
		goto out;
	}

	/* Verify that this is the TPM device */
	if (spi->chip_select != 0) {
		retval = -EINVAL;
		goto err;
	}

	/*
	 * Disable auto chip select for the TPM channel.
	 */
	retval = xapea00x_br_disable_cs(dev, 0);
	if (retval)
		goto err;

	/* De-assert chip select for the TPM channel. */
	retval = xapea00x_br_deassert_cs(dev, 0);
	if (retval)
		goto err;

	goto out;

err:
	dev_err(&dev->interface->dev,
		"configuring SPI channel failed with %d\n", retval);

out:
	mutex_unlock(&dev->usb_mutex);
	return retval;
}

/**
 * xapea00x_spi_transfer - Execute a single SPI transfer.
 * @dev: pointer to the device to do the transfer on
 * @tx_buf: pointer to the data to send, if not NULL
 * @rx_buf: pointer to the buffer to store the received data, if not NULL
 * @len: length in bytes of the data to send/receive
 * @cs_hold: If non-zero, the chip select will remain asserted
 * @delay_usecs: If nonzero, how long to delay after last bit transfer
 *
 * Context: !in_interrupt()
 *
 * Return: If successful, 0. Otherwise a negative error number.
 */
int xapea00x_spi_transfer(struct xapea00x_device *dev,
			  const void *tx_buf, void *rx_buf, u32 len,
			  int cs_hold, u16 delay_usecs)
{
	int retval;

	/* Assert chip select */
	retval = xapea00x_br_assert_cs(dev, 0);
	if (retval)
		goto out;

	/* empty transfer */
	if (!tx_buf && !rx_buf)
		retval = 0;
	/* read transfer */
	else if (!tx_buf)
		retval = xapea00x_br_spi_read(dev, rx_buf, len);
	/* write transfer */
	else if (!rx_buf)
		retval = xapea00x_br_spi_write(dev, tx_buf, len);
	/* write_read transfer */
	else
		retval = xapea00x_br_spi_write_read(dev, tx_buf, rx_buf, len);

	/* Deassert chip select, if requested */
	if (!cs_hold)
		retval = xapea00x_br_deassert_cs(dev, 0);

	/* Delay for the requested time */
	udelay(delay_usecs);

out:
	return retval;
}

/**
 * xapea00x_spi_transfer_one_message - Execute a full SPI message.
 * @master: The SPI master on which to execute the message.
 * @msg: The SPI message to execute.
 *
 * Context: !in_interrupt()
 *
 * Return: If successful, 0. Otherwise a negative erorr number.
 */
static int xapea00x_spi_transfer_one_message(struct spi_master *master,
					     struct spi_message *msg)
{
	struct xapea00x_device *dev;
	struct spi_transfer *xfer;
	int is_last, retval;

	dev = spi_master_get_devdata(master);

	mutex_lock(&dev->usb_mutex);
	if (!dev->interface) {
		retval = -ENODEV;
		goto out;
	}

	/* perform all transfers */
	list_for_each_entry(xfer, &msg->transfers, transfer_list) {
		is_last = list_is_last(&xfer->transfer_list, &msg->transfers);

		/* Transfer message */
		retval = xapea00x_spi_transfer(dev, xfer->tx_buf,
					       xfer->rx_buf, xfer->len,
					       is_last == xfer->cs_change,
					       xfer->delay_usecs);
		if (retval)
			goto out;

		msg->actual_length += xfer->len;
	}

	retval = 0;

out:
	msg->status = retval;
	spi_finalize_current_message(master);

	mutex_unlock(&dev->usb_mutex);
	return retval;
}

/**
 * xapea00x_spi_probe - Register and configure the SPI master.
 * @dev: the device whose SPI master to register
 *
 * Return: If successful, 0. Otherwise a negative error number.
 */
int xapea00x_spi_probe(struct xapea00x_device *dev)
{
	struct spi_master *spi_master;
	int retval;

	spi_master = spi_alloc_master(&dev->udev->dev, sizeof(void *));
	if (!spi_master) {
		retval = -ENOMEM;
		goto err_out;
	}

	spi_master_set_devdata(spi_master, dev);

	spi_master->min_speed_hz = 93 * 1000 + 800; /* 93.9kHz */
	spi_master->max_speed_hz = 12 * 1000 * 1000; /* 12 MHz */

	spi_master->bus_num = -1; /* dynamically assigned */
	spi_master->num_chipselect = 1;
	spi_master->mode_bits = SPI_MODE_0;

	spi_master->flags = 0;
	spi_master->setup = xapea00x_spi_setup;
	spi_master->transfer_one_message = xapea00x_spi_transfer_one_message;

	dev->spi_master = spi_master;
	retval = spi_register_master(dev->spi_master);
	if (retval)
		goto free_spi;

	return 0;

free_spi:
	spi_master_put(spi_master);
	dev->spi_master = NULL;

err_out:
	return retval;
}


xapea00x-core.c
/* XAP-EA-00x driver for Linux
 *
 *  Copyright (c) 2017-2020 Xaptum, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#include "xapea00x.h"
#include "xapea00x-backports.h"

#define XAPEA00X_TPM_MODALIAS	"tpm_tis_spi"

#define kref_to_xapea00x(k) container_of(k, struct xapea00x_device, kref)

static void xapea00x_delete(struct kref *kref)
{
	struct xapea00x_device *dev = kref_to_xapea00x(kref);

	usb_put_dev(dev->udev);
	kfree(dev);
}

struct xapea00x_async_probe {
	struct work_struct work;
	struct xapea00x_device *dev;
};

#define work_to_probe(w) container_of(w, struct xapea00x_async_probe, work)

/**
 * xapea00x_init_async_probe - initialize an async probe with the
 * specified values.
 * @probe: pointer to the async_probe to initialize
 * @dev: pointer to the device to probe
 * @f: pointer to the probe function
 */
static void xapea00x_init_async_probe(struct xapea00x_async_probe *probe,
				      struct xapea00x_device *dev,
				      void (*f)(struct work_struct *work))
{
	INIT_WORK(&probe->work, f);
	probe->dev = dev;

	kref_get(&dev->kref);
	spi_master_get(dev->spi_master);
}

/**
 * xapea00x_cleanup_async_probe - clean up the internals of the async
 * probe. Call this method after the probe has completed.
 *
 * The caller is responsible for freeing the probe itself, if
 * dynamically allocated.
 *
 * @probe: pointer to the async_probe to clean up
 */
static void xapea00x_cleanup_async_probe(struct xapea00x_async_probe *probe)
{
	spi_master_put(probe->dev->spi_master);
	kref_put(&probe->dev->kref, xapea00x_delete);
}

static struct spi_board_info tpm_board_info = {
	.modalias	= XAPEA00X_TPM_MODALIAS,
	.max_speed_hz	= 43 * 1000 * 1000, // Hz
	.chip_select	= 0,
	.mode		= SPI_MODE_0
};

/**
 * xapea00x_tpm_probe - Register and initialize the TPM device
 * @work: the work struct contained by the xapea00x device
 *
 * Context: !in_interrupt()
 */
static void xapea00x_tpm_probe(struct work_struct *work)
{
	struct xapea00x_async_probe *probe = work_to_probe(work);
	struct xapea00x_device *dev = probe->dev;
	struct spi_master *spi_master = dev->spi_master;
	struct spi_device *tpm;
	int retval;

	mutex_lock(&dev->usb_mutex);
	if (!dev->interface) {
		retval = -ENODEV;
		goto out;
	}
	/*
	 * This driver is the "platform" in TPM terminology. Before
	 * passing control of the TPM to the Linux TPM subsystem, do
	 * the TPM initialization normally done by the platform code
	 * (e.g., BIOS).
	 */
	retval = xapea00x_tpm_platform_initialize(dev);
	if (retval) {
		dev_err(&dev->interface->dev,
			"unable to do TPM platform initialization: %d\n",
			retval);
		goto err;
	}

	/*
	 * Now register the TPM with the Linux TPM subsystem.  This
	 * may call through to xapea00x_spi_transfer_one_message(), so
	 * don't hold usb_mutex here.
	 */
	mutex_unlock(&dev->usb_mutex);
	tpm = spi_new_device(spi_master, &tpm_board_info);
	mutex_lock(&dev->usb_mutex);
	if (!dev->interface) {
		retval = -ENODEV;
		goto out;
	}
	if (!tpm) {
		retval = -ENODEV;
		dev_err(&dev->interface->dev,
			"unable to add spi device for TPM\n");
		goto err;
	}

	dev->tpm = tpm;
	goto out;

err:
	dev_err(&dev->interface->dev,
		"TPM initialization failed with %d\n", retval);

out:
	mutex_unlock(&dev->usb_mutex);
	xapea00x_cleanup_async_probe(probe);
	kzfree(probe);
}

/*******************************************************************************
 * USB driver structs and functions
 */

static const struct usb_device_id xapea00x_devices[] = {
	{ USB_DEVICE(USB_VENDOR_ID_SILABS, USB_PRODUCT_ID_XAPEA001) },
	{ USB_DEVICE(USB_VENDOR_ID_XAPTUM, USB_PRODUCT_ID_XAPEA002) },
	{ USB_DEVICE(USB_VENDOR_ID_XAPTUM, USB_PRODUCT_ID_XAPEA003) },
	{ USB_DEVICE(USB_VENDOR_ID_XAPTUM, USB_PRODUCT_ID_XAPEA004) },
	{ }
};
MODULE_DEVICE_TABLE(usb, xapea00x_devices);

static int xapea00x_probe(struct usb_interface *interface,
			  const struct usb_device_id *id)
{
	struct xapea00x_device *dev;
	struct xapea00x_async_probe *probe;
	int retval;

	dev = kzalloc(sizeof(*dev), GFP_KERNEL);
	if (!dev)
		return -ENOMEM;

	kref_init(&dev->kref);
	mutex_init(&dev->usb_mutex);

	/* ---------------------- USB ------------------------ */
	dev->interface = interface;
	dev->udev = usb_get_dev(interface_to_usbdev(interface));

	dev->vid = __le16_to_cpu(dev->udev->descriptor.idVendor);
	dev->pid = __le16_to_cpu(dev->udev->descriptor.idProduct);

	retval = usb_find_common_endpoints(interface->cur_altsetting,
					   &dev->bulk_in, &dev->bulk_out,
					   NULL, NULL);
	if (retval) {
		dev_err(&interface->dev,
			"could not find both bulk-in and bulk-out endpoints\n");
		goto free_dev;
	}

	usb_set_intfdata(interface, dev);

	/* ---------------------- SPI Master ------------------------ */
	retval = xapea00x_spi_probe(dev);
	if (retval) {
		dev_err(&interface->dev, "could not initialize SPI master\n");
		goto free_dev;
	}

	/* ---------------------- TPM SPI Device ------------------------ */
	probe = kzalloc(sizeof(*probe), GFP_KERNEL);
	if (!probe) {
		retval = -ENOMEM;
		goto free_spi;
	}
	xapea00x_init_async_probe(probe, dev, xapea00x_tpm_probe);

	schedule_work(&probe->work);

	/* ---------------------- Finished ------------------------ */
	return 0;

free_spi:
	spi_unregister_master(dev->spi_master);

free_dev:
	kref_put(&dev->kref, xapea00x_delete);

	dev_err(&interface->dev, "device failed with %d\n", retval);
	return retval;
}

static void xapea00x_disconnect(struct usb_interface *interface)
{
	struct xapea00x_device *dev = usb_get_intfdata(interface);

	usb_set_intfdata(interface, NULL);
	spi_unregister_master(dev->spi_master);

	mutex_lock(&dev->usb_mutex);
	dev->interface = NULL;
	mutex_unlock(&dev->usb_mutex);

	kref_put(&dev->kref, xapea00x_delete);
}

static struct usb_driver xapea00x_driver = {
	.name		= KBUILD_MODNAME,
	.probe		= xapea00x_probe,
	.disconnect	= xapea00x_disconnect,
	.id_table	= xapea00x_devices
};

module_usb_driver(xapea00x_driver);

MODULE_AUTHOR("David R. Bild <david.bild@xaptum.com>");
MODULE_DESCRIPTION("Xaptum XAP-EA-00x ENF Access card");
MODULE_LICENSE("GPL");


tpm/tpm_tis_spi.c
/*
 * Copyright (C) 2015 Infineon Technologies AG
 * Copyright (C) 2016 STMicroelectronics SAS
 *
 * Authors:
 * Peter Huewe <peter.huewe@infineon.com>
 * Christophe Ricard <christophe-h.ricard@st.com>
 *
 * Maintained by: <tpmdd-devel@lists.sourceforge.net>
 *
 * Device driver for TCG/TCPA TPM (trusted platform module).
 * Specifications at www.trustedcomputinggroup.org
 *
 * This device driver implements the TPM interface as defined in
 * the TCG TPM Interface Spec version 1.3, revision 27 via _raw/native
 * SPI access_.
 *
 * It is based on the original tpm_tis device driver from Leendert van
 * Dorn and Kyleen Hall and Jarko Sakkinnen.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, version 2 of the
 * License.
 */

#include <linux/init.h>
#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/slab.h>
#include <linux/interrupt.h>
#include <linux/wait.h>
#include <linux/acpi.h>
#include <linux/freezer.h>

#include <linux/spi/spi.h>
#include <linux/gpio.h>
#include <linux/of_irq.h>
#include <linux/of_gpio.h>
#include <linux/tpm.h>
#include "tpm.h"
#include "tpm_tis_core.h"

#define MAX_SPI_FRAMESIZE 64

struct tpm_tis_spi_phy {
	struct tpm_tis_data priv;
	struct spi_device *spi_device;
	u8 *iobuf;
};

static inline struct tpm_tis_spi_phy *to_tpm_tis_spi_phy(struct tpm_tis_data *data)
{
	return container_of(data, struct tpm_tis_spi_phy, priv);
}

static int tpm_tis_spi_transfer(struct tpm_tis_data *data, u32 addr, u16 len,
				u8 *in, const u8 *out)
{
	struct tpm_tis_spi_phy *phy = to_tpm_tis_spi_phy(data);
	int ret = 0;
	int i;
	struct spi_message m;
	struct spi_transfer spi_xfer;
	u8 transfer_len;

	spi_bus_lock(phy->spi_device->master);

	while (len) {
		transfer_len = min_t(u16, len, MAX_SPI_FRAMESIZE);

		phy->iobuf[0] = (in ? 0x80 : 0) | (transfer_len - 1);
		phy->iobuf[1] = 0xd4;
		phy->iobuf[2] = addr >> 8;
		phy->iobuf[3] = addr;

		memset(&spi_xfer, 0, sizeof(spi_xfer));
		spi_xfer.tx_buf = phy->iobuf;
		spi_xfer.rx_buf = phy->iobuf;
		spi_xfer.len = 4;
		spi_xfer.cs_change = 1;

		spi_message_init(&m);
		spi_message_add_tail(&spi_xfer, &m);
		ret = spi_sync_locked(phy->spi_device, &m);
		if (ret < 0)
			goto exit;

		if ((phy->iobuf[3] & 0x01) == 0) {
			// handle SPI wait states
			phy->iobuf[0] = 0;

			for (i = 0; i < TPM_RETRY; i++) {
				spi_xfer.len = 1;
				spi_message_init(&m);
				spi_message_add_tail(&spi_xfer, &m);
				ret = spi_sync_locked(phy->spi_device, &m);
				if (ret < 0)
					goto exit;
				if (phy->iobuf[0] & 0x01)
					break;
			}

			if (i == TPM_RETRY) {
				ret = -ETIMEDOUT;
				goto exit;
			}
		}

		spi_xfer.cs_change = 0;
		spi_xfer.len = transfer_len;
		spi_xfer.delay_usecs = 5;

		if (in) {
			spi_xfer.tx_buf = NULL;
		} else if (out) {
			spi_xfer.rx_buf = NULL;
			memcpy(phy->iobuf, out, transfer_len);
			out += transfer_len;
		}
                        
		spi_message_init(&m);
		spi_message_add_tail(&spi_xfer, &m);
		ret = spi_sync_locked(phy->spi_device, &m);
		if (ret < 0)
			goto exit;

		if (in) {
			memcpy(in, phy->iobuf, transfer_len);
			in += transfer_len;
		}

		len -= transfer_len;
	}

exit:
	spi_bus_unlock(phy->spi_device->master);
	return ret;
}

static int tpm_tis_spi_read_bytes(struct tpm_tis_data *data, u32 addr,
				  u16 len, u8 *result)
{
	return tpm_tis_spi_transfer(data, addr, len, result, NULL);
}

static int tpm_tis_spi_write_bytes(struct tpm_tis_data *data, u32 addr,
				   u16 len, u8 *value)
{
	return tpm_tis_spi_transfer(data, addr, len, NULL, value);
}

static int tpm_tis_spi_read16(struct tpm_tis_data *data, u32 addr, u16 *result)
{
	__le16 result_le;
	int rc;

	rc = data->phy_ops->read_bytes(data, addr, sizeof(u16),
				       (u8 *)&result_le);
	if (!rc)
		*result = le16_to_cpu(result_le);

	return rc;
}

static int tpm_tis_spi_read32(struct tpm_tis_data *data, u32 addr, u32 *result)
{
	__le32 result_le;
	int rc;

	rc = data->phy_ops->read_bytes(data, addr, sizeof(u32),
				       (u8 *)&result_le);
	if (!rc)
		*result = le32_to_cpu(result_le);

	return rc;
}

static int tpm_tis_spi_write32(struct tpm_tis_data *data, u32 addr, u32 value)
{
	__le32 value_le;
	int rc;

	value_le = cpu_to_le32(value);
	rc = data->phy_ops->write_bytes(data, addr, sizeof(u32),
					(u8 *)&value_le);

	return rc;
}

static const struct tpm_tis_phy_ops tpm_spi_phy_ops = {
	.read_bytes = tpm_tis_spi_read_bytes,
	.write_bytes = tpm_tis_spi_write_bytes,
	.read16 = tpm_tis_spi_read16,
	.read32 = tpm_tis_spi_read32,
	.write32 = tpm_tis_spi_write32,
};

static int tpm_tis_spi_probe(struct spi_device *dev)
{
	struct tpm_tis_spi_phy *phy;

	phy = devm_kzalloc(&dev->dev, sizeof(struct tpm_tis_spi_phy),
			   GFP_KERNEL);
	if (!phy)
		return -ENOMEM;

	phy->spi_device = dev;

	phy->iobuf = devm_kmalloc(&dev->dev, MAX_SPI_FRAMESIZE, GFP_KERNEL);
	if (!phy->iobuf)
		return -ENOMEM;

	return tpm_tis_core_init(&dev->dev, &phy->priv, -1, &tpm_spi_phy_ops,
				 NULL);
}

static SIMPLE_DEV_PM_OPS(tpm_tis_pm, tpm_pm_suspend, tpm_tis_resume);

static int tpm_tis_spi_remove(struct spi_device *dev)
{
	struct tpm_chip *chip = spi_get_drvdata(dev);

	tpm_chip_unregister(chip);
	tpm_tis_remove(chip);
	return 0;
}

static const struct spi_device_id tpm_tis_spi_id[] = {
	{"tpm_tis_spi", 0},
	{}
};
MODULE_DEVICE_TABLE(spi, tpm_tis_spi_id);

static const struct of_device_id of_tis_spi_match[] = {
	{ .compatible = "st,st33htpm-spi", },
	{ .compatible = "infineon,slb9670", },
	{ .compatible = "tcg,tpm_tis-spi", },
	{}
};
MODULE_DEVICE_TABLE(of, of_tis_spi_match);

static const struct acpi_device_id acpi_tis_spi_match[] = {
	{"SMO0768", 0},
	{}
};
MODULE_DEVICE_TABLE(acpi, acpi_tis_spi_match);

static struct spi_driver tpm_tis_spi_driver = {
	.driver = {
		.owner = THIS_MODULE,
		.name = "tpm_tis_spi",
		.pm = &tpm_tis_pm,
		.of_match_table = of_match_ptr(of_tis_spi_match),
		.acpi_match_table = ACPI_PTR(acpi_tis_spi_match),
	},
	.probe = tpm_tis_spi_probe,
	.remove = tpm_tis_spi_remove,
	.id_table = tpm_tis_spi_id,
};
module_spi_driver(tpm_tis_spi_driver);

MODULE_DESCRIPTION("TPM Driver for native SPI access");
MODULE_LICENSE("GPL");


xapea00x-backports.c
/*
 * XAP-EA-00x driver for Linux
 *
 *  Copyright (c) 2017 Xaptum, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#include <linux/module.h>

#include "xapea00x-backports.h"

#if LINUX_VERSION_CODE < KERNEL_VERSION(4,12,0)
// Backported from the kernel source
static bool match_endpoint(struct usb_endpoint_descriptor *epd,
		struct usb_endpoint_descriptor **bulk_in,
		struct usb_endpoint_descriptor **bulk_out,
		struct usb_endpoint_descriptor **int_in,
		struct usb_endpoint_descriptor **int_out)
{
	switch (usb_endpoint_type(epd)) {
	case USB_ENDPOINT_XFER_BULK:
		if (usb_endpoint_dir_in(epd)) {
			if (bulk_in && !*bulk_in) {
				*bulk_in = epd;
				break;
			}
		} else {
			if (bulk_out && !*bulk_out) {
				*bulk_out = epd;
				break;
			}
		}

		return false;
	case USB_ENDPOINT_XFER_INT:
		if (usb_endpoint_dir_in(epd)) {
			if (int_in && !*int_in) {
				*int_in = epd;
				break;
			}
		} else {
			if (int_out && !*int_out) {
				*int_out = epd;
				break;
			}
		}

		return false;
	default:
		return false;
	}

	return (!bulk_in || *bulk_in) && (!bulk_out || *bulk_out) &&
			(!int_in || *int_in) && (!int_out || *int_out);
}

/**
 * usb_find_common_endpoints() -- look up common endpoint descriptors
 * @alt:	alternate setting to search
 * @bulk_in:	pointer to descriptor pointer, or NULL
 * @bulk_out:	pointer to descriptor pointer, or NULL
 * @int_in:	pointer to descriptor pointer, or NULL
 * @int_out:	pointer to descriptor pointer, or NULL
 *
 * Search the alternate setting's endpoint descriptors for the first bulk-in,
 * bulk-out, interrupt-in and interrupt-out endpoints and return them in the
 * provided pointers (unless they are NULL).
 *
 * If a requested endpoint is not found, the corresponding pointer is set to
 * NULL.
 *
 * Return: Zero if all requested descriptors were found, or -ENXIO otherwise.
 */
int usb_find_common_endpoints(struct usb_host_interface *alt,
                              struct usb_endpoint_descriptor **bulk_in,
                              struct usb_endpoint_descriptor **bulk_out,
                              struct usb_endpoint_descriptor **int_in,
                              struct usb_endpoint_descriptor **int_out)
{
        struct usb_endpoint_descriptor *epd;
        int i;
        
        if (bulk_in)
                *bulk_in = NULL;
        if (bulk_out)
                *bulk_out = NULL;
        if (int_in)
                *int_in = NULL;
        if (int_out)
                *int_out = NULL;

        for (i = 0; i < alt->desc.bNumEndpoints; ++i) {
                epd = &alt->endpoint[i].desc;

                if (match_endpoint(epd, bulk_in, bulk_out, int_in, int_out))
                        return 0;
        }

        return -ENXIO;
}
#endif

MODULE_LICENSE("GPL");


xapea00x.h
/* XAP-EA-00x driver for Linux
 *
 *  Copyright (c) 2017-2020 Xaptum, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */

#ifndef _XAPEA00X_H
#define _XAPEA00X_H

#include <linux/kernel.h>
#include <linux/kref.h>
#include <linux/module.h>
#include <linux/mutex.h>
#include <linux/slab.h>
#include <linux/spi/spi.h>
#include <linux/usb.h>
#include <linux/workqueue.h>

#define USB_VENDOR_ID_SILABS           0x10c4
#define USB_VENDOR_ID_XAPTUM           0x2FE0

#define USB_PRODUCT_ID_XAPEA001        0x8BDE
#define USB_PRODUCT_ID_XAPEA002        0x8BDE
#define USB_PRODUCT_ID_XAPEA003        0x8BEE
#define USB_PRODUCT_ID_XAPEA004        0x8BFE

struct xapea00x_device {
	struct kref kref;

	struct usb_device *udev;
	/*
	 * The interface pointer will be set NULL when the device
	 * disconnects.  Accessing it safe only while holding the
	 * usb_mutex.
	 */
	struct usb_interface *interface;
	/*
	 * Th usb_mutex must be held while synchronous USB requests are
	 * in progress. It is acquired during disconnect to be sure
	 * that there is not an outstanding request.
	 */
	struct mutex usb_mutex;

	struct usb_endpoint_descriptor *bulk_in;
	struct usb_endpoint_descriptor *bulk_out;

	u16 pid;
	u16 vid;

	struct spi_master *spi_master;
	struct spi_device *tpm;
};

/* Public bridge functions */
int xapea00x_br_disable_cs(struct xapea00x_device *dev, u8 channel);
int xapea00x_br_assert_cs(struct xapea00x_device *dev, u8 channel);
int xapea00x_br_deassert_cs(struct xapea00x_device *dev, u8 channel);

int xapea00x_br_spi_read(struct xapea00x_device *dev, void *rx_buf, int len);
int xapea00x_br_spi_write(struct xapea00x_device *dev, const void *tx_buf,
			  int len);
int xapea00x_br_spi_write_read(struct xapea00x_device *dev, const void *tx_buf,
			       void *rx_buf, int len);

/* Shared SPI function */
int xapea00x_spi_probe(struct xapea00x_device *dev);
int xapea00x_spi_transfer(struct xapea00x_device *dev,
			  const void *tx_buf, void *rx_buf, u32 len,
			  int cs_hold, u16 delay_usecs);

/* Shared TPM functions */
int xapea00x_tpm_platform_initialize(struct xapea00x_device *dev);

#endif /* _XAPEA00X_H */
